File: apps/loomx-api/src/routes/sql.ts
import { Router, type IRouter } from 'express';
import { asyncHandler, ValidationError } from '../middleware/errorHandler';
import { DatasetsService } from '../services/datasets.service';
import { pythonProxyService } from '../services/pythonProxy.service';
import { QueryHistoryService } from '../services/queryHistory.service';
import { getCurrentUserId } from '../middleware/userContext';
import { buildChartPreviewQuery, buildDistinctFilterValuesQuery } from '../services/queryGenerator.service';
import { extractTablesFromSql } from '../services/sqlTableExtractor';

/**
 * Maps a raw source string coming from the frontend into a canonical
 * trigger_source value that is stored in query_history.
 *
 * Canonical vocabulary:
 *   lab                   – SQL Lab editor (ad-hoc)
 *   dataset-preview       – Dataset detail page preview
 *   dataset-filter        – Dataset builder filter-value lookup
 *   chart-builder         – Chart builder page — main query
 *   chart-builder-filter  – Chart builder page — filter dropdown values
 *   dashboard-chart       – Chart rendering inside a dashboard
 *   dashboard-filter      – Dashboard filter bar value lookup
 */
function canonicalSource(raw: string | undefined | null): string {
  switch ((raw || '').toLowerCase()) {
    case 'dashboard':
    case 'dashboard-chart':
      return 'dashboard-chart';
    case 'chart-builder':
    case 'chart-builder-filter':
      return raw!.toLowerCase();
    case 'dashboard-filter':
    case 'dataset-preview':
    case 'dataset-filter':
    case 'lab':
      return raw!.toLowerCase();
    default:
      return raw || 'chart-builder';
  }
}

const router: IRouter = Router();
const datasetsService = new DatasetsService();
const queryHistoryService = new QueryHistoryService();


/**
 * POST /api/v1/sql/generate
 * Generate SQL from chart configuration
 */
router.post('/generate', asyncHandler(async (req, res) => {
  const { dataset_id, chart_type, config } = req.body;

  if (!dataset_id || !chart_type || !config) {
    throw new ValidationError('dataset_id, chart_type, and config are required');
  }

  // Load dataset with full details
  const dataset = await datasetsService.getById(String(dataset_id));
  if (!dataset) {
    throw new ValidationError('Dataset not found');
  }

  // Build dimensions payload matching FabricExplorer structure
  // Each dimension needs:
  // - table: dimension table name (e.g., "dbo.DimProduct")
  // - factKey: fact-side join key (e.g., "FactSales.ProductKey")
  // - dimKey: dimension-side join key (e.g., "DimProduct.ProductKey")
  const dimensions = (dataset.dimensions || []).map((dim: any) => ({
    table: dim.dimension_table || dim.table_name,
    factKey: dim.fact_key,  // Fact-side join key
    dimKey: dim.join_key,   // Dimension-side join key
    semanticColumns: []
  }));

  // Build datasource name
  const datasource = dataset.schema_name
    ? `${dataset.schema_name}.${dataset.table_name}`
    : dataset.table_name;

  // Build query parameters
  const params = {
    datasource,
    ...config,
    dimensions,
    columns: dataset.columns || [],  // Pass columns for unqualified name resolution
    database_name: dataset.database_name
  };

  // Generate SQL
  const sqlText = buildChartPreviewQuery(params);

  if (!sqlText) {
    console.error('[SQL Generate] Failed to generate SQL', {
      datasource,
      dimensions: dimensions.length,
      config: JSON.stringify(config).substring(0, 200)
    });
    throw new ValidationError('Failed to generate SQL from configuration');
  }

  // Log generated SQL for debugging
  console.log('[SQL Generate] Generated SQL:', {
    datasource,
    dimensionCount: dimensions.length,
    sqlPreview: sqlText.substring(0, 300) + (sqlText.length > 300 ? '...' : '')
  });

  // Extract tables used
  const tablesUsed = [dataset.table_name];
  if (dimensions && dimensions.length > 0) {
    dimensions.forEach((d: any) => {
      if (d.table && !tablesUsed.includes(d.table)) {
        tablesUsed.push(d.table);
      }
    });
  }

  res.json({
    sql_text: sqlText,
    tables_used: tablesUsed,
    warnings: []
  });
}));

/**
 * GET /api/v1/sql/distinct-filter-values
 * Get distinct values for filter dropdown
 * LIVE DATA - No caching
 *
 * Optional query params for history context:
 *   source       – caller context ('chart-builder-filter', 'dashboard-filter')
 *   chart_id     – chart being configured/rendered
 *   dashboard_id – dashboard (when source is 'dashboard-filter')
 */
router.get('/distinct-filter-values', asyncHandler(async (req, res) => {
  // Disable browser caching - LIVE data only
  res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
  res.setHeader('Pragma', 'no-cache');
  res.setHeader('Expires', '0');

  const { dataset_id, column, fact_key, limit, source, chart_id, dashboard_id } = req.query;

  if (!dataset_id || !column) {
    throw new ValidationError('dataset_id and column are required');
  }

  const rowLimit = limit ? Math.min(parseInt(limit as string, 10), 500) : 100;

  // Load dataset
  const dataset = await datasetsService.getById(String(dataset_id));
  if (!dataset) {
    throw new ValidationError('Dataset not found');
  }

  if (!dataset.table_name) {
    throw new ValidationError('Dataset is missing table_name');
  }

  // Build datasource name
  const datasource = dataset.schema_name
    ? `${dataset.schema_name}.${dataset.table_name}`
    : dataset.table_name;

  // Build dimensions payload.
  // When fact_key is provided, sort the matching dimension to the front so that
  // determineFilteringStrategy picks the correct factKey for role-playing dims
  // (e.g. IDEASBoolFlag used for both IsMSFTTenant and IsStrategicCustomer).
  const hintFactKey = fact_key ? String(fact_key).toLowerCase() : null;
  const rawDimensions = (dataset.dimensions || []).map((dim: any) => ({
    table: dim.dimension_table || dim.table_name,
    factKey: dim.fact_key,
    dimKey: dim.join_key
  }));
  const dimensions = hintFactKey
    ? [...rawDimensions].sort((a, b) => {
        const aMatch = (a.factKey || '').toLowerCase() === hintFactKey ? -1 : 0;
        const bMatch = (b.factKey || '').toLowerCase() === hintFactKey ? -1 : 0;
        return aMatch - bMatch;
      })
    : rawDimensions;

  // Generate query with metadata for optimized filtering
  const queryResult = buildDistinctFilterValuesQuery({
    datasource,
    column: column as string,
    dimensions,
    columns: dataset.columns || [],
    limit: rowLimit
  });

  if (!queryResult) {
    throw new ValidationError('Failed to generate distinct values query');
  }

  const { sql: sqlText, keyColumn, filteringTier } = queryResult;

  // Execute query
  if (!dataset.database_name) {
    throw new ValidationError('Dataset database_name is missing. Please update the dataset configuration.');
  }

  const userId = getCurrentUserId(req);
  const filterTriggerSource = canonicalSource(
    source ? String(source) : 'chart-builder-filter'
  ) || 'chart-builder-filter';

  // Build tables_used from the dataset: fact table + any dimension tables.
  const filterTablesUsed: string[] = [
    dataset.schema_name
      ? `${dataset.schema_name}.${dataset.table_name}`
      : dataset.table_name,
  ];
  (dataset.dimensions || []).forEach((dim: any) => {
    const dimTable = dim.dimension_table || dim.table_name;
    if (dimTable && !filterTablesUsed.includes(dimTable)) {
      filterTablesUsed.push(dimTable);
    }
  });

  const filterRunContext = JSON.stringify({
    source: filterTriggerSource,
    datasetId:     Number(dataset_id),
    column:        column as string,
    filteringTier,
    database:      dataset.database_name,
    ...(chart_id     ? { chartId:     Number(chart_id)     } : {}),
    ...(dashboard_id ? { dashboardId: Number(dashboard_id) } : {}),
  });

  const filterStartTime = Date.now();
  let filterResult: Awaited<ReturnType<typeof pythonProxyService.executeQuery>>;
  try {
    filterResult = await pythonProxyService.executeQuery(sqlText, dataset.database_name);
  } catch (error) {
    // Log the failed filter-values query before re-throwing.
    try {
      await queryHistoryService.create({
        sql_text: sqlText,
        duration_ms: Date.now() - filterStartTime,
        row_count: 0,
        status: 'error',
        error_message: error instanceof Error ? error.message : 'Unknown error',
        trigger_source: filterTriggerSource,
        tables_used: JSON.stringify(filterTablesUsed),
        run_context: filterRunContext,
        dataset_id: Number(dataset_id),
        started_at: filterStartTime,
      }, userId);
    } catch (logErr) {
      console.error('[SQL] Failed to log filter-values error to history:', logErr);
    }
    throw error;
  }

  const filterDurationMs = Date.now() - filterStartTime;

  // Log successful filter-values query to history.
  try {
    await queryHistoryService.create({
      sql_text: sqlText,
      duration_ms: filterDurationMs,
      row_count: filterResult.rows?.length || 0,
      status: 'success',
      trigger_source: filterTriggerSource,
      tables_used: JSON.stringify(filterTablesUsed),
      run_context: filterRunContext,
      dataset_id: Number(dataset_id),
      started_at: filterStartTime,
    }, userId);
  } catch (logErr) {
    console.error('[SQL] Failed to log filter-values query to history:', logErr);
  }

  // Transform rows to key-value pairs
  const values = (filterResult.rows || []).map((row: any) => {
    if (Array.isArray(row)) {
      return { key: row[0], value: row[1] || row[0] };
    }
    return { key: row.key || row[0], value: row.value || row[1] || row[0] };
  });

  // Return values with filtering metadata
  res.json({
    success: true,
    values,
    keyColumn,      // fact_key column for tier 1/2 optimization (null if not applicable)
    filteringTier,  // 1 = fastest (fact_key only), 2 = medium (dim key), 3 = slowest (dim display)
  });
}));

/**
 * POST /api/v1/sql/execute
 * Execute arbitrary SQL query
 * LIVE DATA - No caching
 *
 * Optional body fields for richer history context:
 *   source       – caller context ('chart-builder', 'dashboard', etc.)
 *   chart_id     – ID of the chart being rendered
 *   dashboard_id – ID of the dashboard (when source is 'dashboard')
 *   chart_type   – chart template type ('bar', 'line', etc.)
 *   dataset_id   – dataset the query was built from
 *   tables_used  – JSON-stringified array of table names; if omitted,
 *                  extracted server-side from the SQL text
 */
router.post('/execute', asyncHandler(async (req, res) => {
  // Disable browser caching - LIVE data only
  res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
  res.setHeader('Pragma', 'no-cache');
  res.setHeader('Expires', '0');

  const userId = getCurrentUserId(req);
  const {
    sql_text,
    row_limit,
    source,
    tables_used,
    database,
    chart_id,
    dashboard_id,
    chart_type,
    dataset_id,
  } = req.body;

  if (!sql_text || typeof sql_text !== 'string') {
    throw new ValidationError('sql_text is required');
  }

  if (!database) {
    throw new ValidationError('database parameter is required');
  }

  const trigger_source = canonicalSource(source);

  // Resolve tables_used: prefer what the frontend provides, fall back to
  // server-side extraction so the column is never null.
  const resolvedTablesUsed: string = tables_used ||
    JSON.stringify(extractTablesFromSql(sql_text));

  // Consistent run_context: include all available caller metadata.
  const run_context = JSON.stringify({
    source: trigger_source,
    ...(chart_id     != null ? { chartId:     Number(chart_id)     } : {}),
    ...(dashboard_id != null ? { dashboardId: Number(dashboard_id) } : {}),
    ...(chart_type   != null ? { chartType:   chart_type           } : {}),
    ...(dataset_id   != null ? { datasetId:   Number(dataset_id)   } : {}),
    database,
  });

  console.log('[SQL] Executing query on database:', database, '| source:', trigger_source);
  console.log('[SQL] Query preview:', sql_text.substring(0, 100) + (sql_text.length > 100 ? '...' : ''));

  const startTime = Date.now();
  let result;

  try {
    // Python proxy doesn't handle limit - apply it after execution if needed
    result = await pythonProxyService.executeQuery(sql_text, database);

    // Apply row limit if specified
    const limit = row_limit ? Math.min(Math.max(1, parseInt(row_limit, 10)), 5000) : undefined;
    if (limit && result.rows && result.rows.length > limit) {
      result.rows = result.rows.slice(0, limit);
      result.rowCount = limit;
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error';
    console.error('[SQL] Query execution failed:', errorMessage);

    try {
      await queryHistoryService.create({
        sql_text,
        duration_ms: Date.now() - startTime,
        row_count: 0,
        status: 'error',
        error_message: errorMessage,
        trigger_source,
        tables_used: resolvedTablesUsed,
        run_context,
        dataset_id: dataset_id ? Number(dataset_id) : null,
        started_at: startTime,
      }, userId);
    } catch (err) {
      console.error('[SQL] Failed to log error to query history:', err);
    }

    throw error;
  }

  const durationMs = Date.now() - startTime;

  try {
    await queryHistoryService.create({
      sql_text,
      duration_ms: durationMs,
      row_count: result.rowCount || 0,
      status: 'success',
      trigger_source,
      tables_used: resolvedTablesUsed,
      run_context,
      dataset_id: dataset_id ? Number(dataset_id) : null,
      started_at: startTime,
    }, userId);
  } catch (err) {
    console.error('[SQL] Failed to log query history:', err);
  }

  res.json({
    columns: result.columns || [],
    rows: result.rows || [],
    message: result.rowCount ? `Returned ${result.rowCount} rows` : undefined,
    duration_ms: durationMs,
  });
}));

export default router;

File: apps/loomx-api/routers/sql.py
"""SQL router — /api/v1/sql."""

import json
import time
from fastapi import APIRouter, Request, Response, HTTPException, Query, Depends
from middleware.auth import require_auth
import services.datasets as datasets_svc
import services.query_history as history_svc
from services.query_generator import build_chart_preview_query, build_distinct_filter_values_query
from services.sql_table_extractor import extract_tables_from_sql
import database.pool as pool

router = APIRouter()
NO_CACHE = {"Cache-Control": "no-cache, no-store, must-revalidate", "Pragma": "no-cache", "Expires": "0"}

_CANONICAL_SOURCE = {
    "dashboard": "dashboard-chart",
    "dashboard-chart": "dashboard-chart",
    "chart-builder": "chart-builder",
    "chart-builder-filter": "chart-builder-filter",
    "dashboard-filter": "dashboard-filter",
    "dataset-preview": "dataset-preview",
    "dataset-filter": "dataset-filter",
    "lab": "lab",
}


def canonical_source(raw: str) -> str:
    return _CANONICAL_SOURCE.get((raw or "").lower(), raw or "chart-builder")


@router.post("/sql/generate")
def generate_sql(data: dict, user: str = Depends(require_auth)):
    dataset_id = data.get("dataset_id")
    chart_type = data.get("chart_type")
    config = data.get("config")

    if not dataset_id or not chart_type or not config:
        raise HTTPException(status_code=400, detail="dataset_id, chart_type, and config are required")

    dataset = datasets_svc.get_dataset_by_id(str(dataset_id))
    if not dataset:
        raise HTTPException(status_code=400, detail="Dataset not found")

    dimensions = [
        {
            "table": dim.get("dimension_table") or dim.get("table_name"),
            "factKey": dim.get("fact_key"),
            "dimKey": dim.get("join_key"),
            "semanticColumns": [],
        }
        for dim in (dataset.get("dimensions") or [])
    ]

    datasource = (
        f"{dataset['schema_name']}.{dataset['table_name']}"
        if dataset.get("schema_name")
        else dataset.get("table_name") or ""
    )

    params = {
        "datasource": datasource,
        **config,
        "dimensions": dimensions,
        "columns": dataset.get("columns") or [],
        "database_name": dataset.get("database_name"),
    }

    sql_text = build_chart_preview_query(params)
    if not sql_text:
        raise HTTPException(status_code=400, detail="Failed to generate SQL from configuration")

    tables_used = [dataset.get("table_name") or datasource]
    for d in dimensions:
        if d.get("table") and d["table"] not in tables_used:
            tables_used.append(d["table"])

    return {"sql_text": sql_text, "tables_used": tables_used, "warnings": []}


@router.get("/sql/distinct-filter-values")
def distinct_filter_values(
    response: Response,
    dataset_id: str = Query(...), column: str = Query(...),
    fact_key: str = Query(default=None),
    limit: int = Query(default=100),
    source: str = Query(default=None),
    chart_id: str = Query(default=None),
    dashboard_id: str = Query(default=None),
    user: str = Depends(require_auth),
):
    response.headers.update(NO_CACHE)
    user_id = user
    row_limit = min(limit, 500)

    dataset = datasets_svc.get_dataset_by_id(dataset_id)
    if not dataset:
        raise HTTPException(status_code=400, detail="Dataset not found")
    if not dataset.get("table_name"):
        raise HTTPException(status_code=400, detail="Dataset is missing table_name")
    if not dataset.get("database_name"):
        raise HTTPException(status_code=400, detail="Dataset database_name is missing")

    datasource = (
        f"{dataset['schema_name']}.{dataset['table_name']}"
        if dataset.get("schema_name")
        else dataset["table_name"]
    )

    hint_fk = fact_key.lower() if fact_key else None
    raw_dims = [
        {"table": dim.get("dimension_table") or dim.get("table_name"),
         "factKey": dim.get("fact_key"), "dimKey": dim.get("join_key")}
        for dim in (dataset.get("dimensions") or [])
    ]
    if hint_fk:
        raw_dims.sort(key=lambda d: 0 if (d.get("factKey") or "").lower() == hint_fk else 1)

    query_result = build_distinct_filter_values_query({
        "datasource": datasource, "column": column,
        "dimensions": raw_dims, "columns": dataset.get("columns") or [],
        "limit": row_limit,
    })
    if not query_result:
        raise HTTPException(status_code=400, detail="Failed to generate distinct values query")

    sql_text = query_result["sql"]
    key_column = query_result["keyColumn"]
    filtering_tier = query_result["filteringTier"]

    filter_trigger = canonical_source(source or "chart-builder-filter") or "chart-builder-filter"
    tables_used_list = [datasource]
    for dim in (dataset.get("dimensions") or []):
        dt = dim.get("dimension_table") or dim.get("table_name")
        if dt and dt not in tables_used_list:
            tables_used_list.append(dt)

    run_context = json.dumps({
        "source": filter_trigger, "datasetId": int(dataset_id), "column": column,
        "filteringTier": filtering_tier, "database": dataset["database_name"],
        **({} if not chart_id else {"chartId": int(chart_id)}),
        **({} if not dashboard_id else {"dashboardId": int(dashboard_id)}),
    })

    start_time = int(time.time() * 1000)
    try:
        result = pool.execute_query(sql_text, dataset["database_name"])
    except Exception as e:
        duration_ms = int(time.time() * 1000) - start_time
        try:
            history_svc.create_history({
                "sql_text": sql_text, "duration_ms": duration_ms,
                "row_count": 0, "status": "error", "error_message": str(e),
                "trigger_source": filter_trigger,
                "tables_used": json.dumps(tables_used_list),
                "run_context": run_context, "dataset_id": int(dataset_id),
                "started_at": start_time,
            }, user_id)
        except Exception:
            pass
        raise HTTPException(status_code=500, detail="Query execution failed")

    duration_ms = int(time.time() * 1000) - start_time
    try:
        history_svc.create_history({
            "sql_text": sql_text, "duration_ms": duration_ms,
            "row_count": len(result.get("rows") or []), "status": "success",
            "trigger_source": filter_trigger,
            "tables_used": json.dumps(tables_used_list),
            "run_context": run_context, "dataset_id": int(dataset_id),
            "started_at": start_time,
        }, user_id)
    except Exception:
        pass

    rows = result.get("rows_objects") or result.get("rows") or []
    values = []
    for row in rows:
        if isinstance(row, list):
            values.append({"key": row[0], "value": row[1] if len(row) > 1 else row[0]})
        elif isinstance(row, dict):
            values.append({"key": row.get("key", row.get("value")), "value": row.get("value", row.get("key"))})
        else:
            values.append({"key": row, "value": row})

    return {"success": True, "values": values, "keyColumn": key_column, "filteringTier": filtering_tier}


@router.post("/sql/execute")
def execute_sql(data: dict, response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    user_id = user

    sql_text = data.get("sql_text") or ""
    database = data.get("database") or ""
    source = data.get("source")
    tables_used = data.get("tables_used")
    chart_id = data.get("chart_id")
    dashboard_id = data.get("dashboard_id")
    chart_type = data.get("chart_type")
    dataset_id = data.get("dataset_id")
    row_limit = data.get("row_limit")

    if not sql_text:
        raise HTTPException(status_code=400, detail="sql_text is required")
    if not database:
        raise HTTPException(status_code=400, detail="database parameter is required")

    trigger_source = canonical_source(source)
    resolved_tables = tables_used or json.dumps(extract_tables_from_sql(sql_text))
    run_context = json.dumps({
        "source": trigger_source, "database": database,
        **({} if chart_id is None else {"chartId": int(chart_id)}),
        **({} if dashboard_id is None else {"dashboardId": int(dashboard_id)}),
        **({} if chart_type is None else {"chartType": chart_type}),
        **({} if dataset_id is None else {"datasetId": int(dataset_id)}),
    })

    start_time = int(time.time() * 1000)
    try:
        result = pool.execute_query(sql_text, database)

        if row_limit:
            lim = min(max(1, int(row_limit)), 5000)
            rows = result.get("rows") or []
            if len(rows) > lim:
                result["rows"] = rows[:lim]
                result["row_count"] = lim
    except Exception as e:
        duration_ms = int(time.time() * 1000) - start_time
        try:
            history_svc.create_history({
                "sql_text": sql_text, "duration_ms": duration_ms,
                "row_count": 0, "status": "error", "error_message": str(e),
                "trigger_source": trigger_source, "tables_used": resolved_tables,
                "run_context": run_context,
                "dataset_id": int(dataset_id) if dataset_id is not None else None,
                "started_at": start_time,
            }, user_id)
        except Exception:
            pass
        raise HTTPException(status_code=500, detail="Query execution failed")

    duration_ms = int(time.time() * 1000) - start_time
    try:
        history_svc.create_history({
            "sql_text": sql_text, "duration_ms": duration_ms,
            "row_count": result.get("row_count") or 0, "status": "success",
            "trigger_source": trigger_source, "tables_used": resolved_tables,
            "run_context": run_context,
            "dataset_id": int(dataset_id) if dataset_id is not None else None,
            "started_at": start_time,
        }, user_id)
    except Exception:
        pass

    row_count = result.get("row_count") or 0
    return {
        "columns": result.get("columns") or [],
        "rows": result.get("rows") or [],
        "message": f"Returned {row_count} rows" if row_count else None,
        "duration_ms": duration_ms,
    }
